README.md
# NG OIDC Client

[![npm](https://img.shields.io/npm/v/ng-oidc-client.svg)](https://www.npmjs.com/package/ng-oidc-client)
[![npm](https://img.shields.io/npm/dm/ng-oidc-client.svg)](https://www.npmjs.com/ng-oidc-client)
[![Maintenance](https://img.shields.io/maintenance/yes/2018.svg)](https://www.npmjs.com/ng-oidc-client)

An Angular package wrapping [oidc-client-js][1] to manage authentication with OIDC and OAuth2 in a reactive way using [NgRx][2].

<img src="https://github.com/sourcenetwork/ng-oidc-client/blob/master/images/oidc-state-devtools-animated.gif" alt="OIDC State in Redux DevTools">

## Getting started 🚀
The configuration for the examples are based on running IdentityServer4 on localhost. A ready-to-go reference implementation for testing purposes can be found at [ng-oidc-client-server][4]. 

Install the package
```bash
npm install -s ng-oidc-client
```
if not already the case, install the necessary peer dependencies
```bash
npm install -s @ngrx/store 
npm install -s @ngrx/effects 
npm install -s oidc-client
```

Add the NgOidcClientModule to your AppModule
```diff
export interface State {
  router: RouterReducerState;
}
export const rootStore: ActionReducerMap<State> = {
  router: routerReducer
};

@NgModule({
  declarations: [AppComponent, ProtectedComponent, HomeComponent, LoginComponent],
  imports: [
    BrowserModule,
    StoreModule.forRoot(rootStore),
    EffectsModule.forRoot([]),
+    NgOidcClientModule.forRoot({
+     oidc_config: {
+       authority: 'https://localhost:5001',
+       client_id: 'ng-oidc-client-identity',
+       redirect_uri: 'http://localhost:4200/callback.html',
+       response_type: 'id_token token',
+       scope: 'openid profile offline_access api1',
+       post_logout_redirect_uri: 'http://localhost:4200/signout-callback.html',
+       silent_redirect_uri: 'http://localhost:4200/renew-callback.html',
+       accessTokenExpiringNotificationTime: 10,
+       automaticSilentRenew: true,
+       userStore: new WebStorageStateStore({ store: window.localStorage })
+     },
+     log: {
+       logger: console,
+       level: Log.NONE
+     }
+   })
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule {}
```
For a complete list of configuration options please see the [official oidc-client-js documentation][3].

**Coming soon: Example configuration Using ng-oidc-client with Auth0 and Okta**

Inject the `OidcFacade` in your Component
```typescript
...
export class HomeComponent {
  constructor(private oidcFacade: OidcFacade) {
    // Call to get user from storage
    this.oidcFacade.getOidcUser();
  }

  loginPopup() {
    this.oidcFacade.signinPopup();
  }

  logoutPopup() {
    this.oidcFacade.signoutPopup();
  }
}
```
Alternatively you can you also use `.signinRedirect();` or `.signoutRedirect();`. 

Create a new directory called `static` below the `src` folder, to serve the static callback HTML sites.
```bash
src/static
├── callback.html
├── renew-callback.html
└── signout-callback.html
```
Example for a `callback.html`
```html
<!DOCTYPE html>
<html>

<head>
  <meta charset="utf-8" />
  <title>Callback</title>
  <link rel="icon"
        type="image/x-icon"
        href="favicon.png">
  <script src="oidc-client.min.js"
          type="application/javascript"></script>
</head>

<body>
  <script>
    var Oidc = window.Oidc;

    var config = {
      userStore: new Oidc.WebStorageStateStore({ store: window.localStorage })
    }

    if ((Oidc && Oidc.Log && Oidc.Log.logger)) {
      Oidc.Log.logger = console;
    }
    var isPopupCallback = JSON.parse(window.localStorage.getItem('ngoidc:isPopupCallback'));

    if (isPopupCallback) {
      new Oidc.UserManager(config).signinPopupCallback();
    } else {
      new Oidc.UserManager(config).signinRedirectCallback().then(t => {
        window.location.href = '/';
      });
    }
  </script>
</body>

</html>
```

Example for a `renew-callback.html`
```html
<!DOCTYPE html>
<html>

<head>
    <meta charset="utf-8" />
    <title>Renew Callback</title>
    <link rel="icon"
          type="image/x-icon"
          href="favicon.png">
</head>

<body>
    <script src="oidc-client.min.js"></script>
    <script>
        var config = {
            userStore: new Oidc.WebStorageStateStore({ store: window.localStorage })
        }
        new Oidc.UserManager(config).signinSilentCallback().catch(function (e) {
            console.error(e);
        });
    </script>
</body>

</html>
```

Example for a `signout-callback.html`
```html
<!DOCTYPE html>
<html>

<head>
  <meta charset="utf-8" />
  <title>Signout Callback</title>
  <link rel="icon"
        type="image/x-icon"
        href="favicon.png">
  <script src="oidc-client.min.js"
          type="application/javascript"></script>
</head>

<body>
  <script>
    var Oidc = window.Oidc;

    var config = {
      userStore: new Oidc.WebStorageStateStore({ store: window.localStorage })
    }

    if ((Oidc && Oidc.Log && Oidc.Log.logger)) {
      Oidc.Log.logger = console;
    }

    var isPopupCallback = JSON.parse(window.localStorage.getItem('ngoidc:isPopupCallback'));

    if (isPopupCallback) {
      new Oidc.UserManager(config).signoutPopupCallback();
    } else {
      new Oidc.UserManager(config).signoutRedirectCallback().then(test => {
        window.location.href = '/';
      });
    }
  </script>
</body>

</html>
```

Modify your `angular.json` to include the `static` assets and `oidc-client`
```diff
...
"assets": [
    "src/favicon.ico",
    "src/assets",
+   {
+     "glob": "**/*",
+     "input": "src/static",
+     "output": "/"
+   },
+   {
+     "glob": "oidc-client.min.js",
+     "input": "node_modules/oidc-client/dist",
+     "output": "/"
+   }
  ],
...
```

You will be able to authenticate against your configurated identity provider and the obtained user will be accessible in through the created state.

## Protecting Routes using an AuthGuard 💂
Create a new Service and implement the AuthGuard interface
```typescript
export class OidcGuardService implements CanActivate {
  constructor(private router: Router, private oidcFacade: OidcFacade) {}

  public canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<boolean> | boolean {
    return this.oidcFacade.identity$.pipe(
      take(1),
      switchMap(user => {
        console.log('Auth Guard - Checking if user exists', user);
        console.log('Auth Guard - Checking if user is expired:', user && user.expired);
        if (user && !user.expired) {
          return of(true);
        } else {
          this.router.navigate(['/login']);
          return of(false);
        }
      })
    );
  }
}
```
The guard needs to inject the `OidcFacade` to check if the user exists and is not expired. It is up to the application flow if the unauthenticated user is redirected to a login route or not.

Add the Guard to the list of providers in your AppModule
```diff
...
providers: [
+   OidcGuardService,
  ],
...
```

## Using HTTP Interceptor to add the Bearer token to API calls 🐻
In case an API requires a valid access token to interact with it, an HTTP Interceptor can be used to add the required Bearer token to each call. 

Create a new Service and implement the HttpInterceptor interface
```typescript
export class OidcInterceptorService implements HttpInterceptor {
  static OidcInterceptorService: any;
  constructor(private oidcFacade: OidcFacade) {}

  intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    return this.oidcFacade.identity$.pipe(
      switchMap(user => {
        if (user && !user.expired && user.access_token) {
          req = req.clone({
            setHeaders: {
              Authorization: `Bearer ${user.access_token}`
            }
          });
        }
        return next.handle(req);
      })
    );
  }
}
```
The interceptor needs to inject the `OidcFacade` to check if the user exists, is not expired and has an access token. The outgoing request will be cloned and an Authorization Header will be added to the request. Additionally, the requests should be filtered to only add the Bearer token where it is needed.

Add the Interceptor to the list of providers in your AppModule
```diff
...
providers: [
+   {
+     provide: HTTP_INTERCEPTORS,
+     useClass: OidcInterceptorService,
+     multi: true
+   }
  ],
...
```

# Docs
You'll find a more detailed [Documentation in our Wiki][5]
## Articles
- [Securing your Angular 7+ application with OIDC and NgRx](https://medium.com/@alx_andru/securing-your-angular-7-application-with-oidc-and-ngrx-7775e8325b51)
  - [Using HttpInterceptor to add a Bearer 🐻 token to API calls with ng-oidc-client](https://medium.com/@alx_andru/using-httpinterceptor-to-add-a-bearer-token-to-api-calls-with-ng-oidc-client-6b088d680e73)
  - [Route Guards💂 to guard routes with ng-oidc-client](https://medium.com/@alx_andru/route-guards-to-guard-routes-with-ng-oidc-client-6a61e6029424)
  - [Ready, Set, OidcActions! 🎬](https://medium.com/@alx_andru/ready-set-oidcactions-a14a3f89cb7f)


# Examples
Coming soon

# Licence
This software is [licensed][6] under the MIT


[1]: https://github.com/IdentityModel/oidc-client-js
[2]: https://github.com/ngrx/platform
[3]: https://github.com/IdentityModel/oidc-client-js/wiki#configuration
[4]: https://github.com/sourcenetwork/ng-oidc-client-server
[5]: https://github.com/sourcenetwork/ng-oidc-client/wiki
[6]: https://github.com/sourcenetwork/ng-oidc-client/blob/master/LICENSE


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.0.4] - 2020-05-26
- Updated to Angular 9
- Restore facade functions

## [2.0.3] - 2020-02-06
## [2.0.2] - 2020-02-06
### Fixed
- Injection Error

## [2.0.1] - 2020-02-04
### Fixes 
- Type error

## [2.0.0] - 2020-02-04
### Breaking Changes
- Updated to Angular/NgRx 8
### Added
- Update packages
- Add configuration options
- Expose oidc client service
- Implement action and effect creators
### Fixed
- Serialize identity
- Added additional request params

## [1.0.7] - 2019-10-27
### Fixed
- Revert to angular 7

## [1.0.6] - 2019-08-13
### Fixed
- Upgrade packages

## [1.0.5] - 2018-12-05
### Added
- Upgrade to Angular 7
### Fixed
- Removed log messages

## [1.0.4] - 2018-10-18
### Fixed
- Fix issue with AoT builds of importing project

## [1.0.3] - 2018-10-18
### Fixed
- Resolve dependencies for OidcService
- Fix Redirect Logout button in example app

## [1.0.2] - 2018-10-18
### Fixed
- Issue [#2](https://github.com/Fileless/ng-oidc-client/issues/2) to enable builds with `--prod` flag

## [1.0.1] - 2018-10-15
### Added
- New configuration `useCallbackFlag?:boolean` to indicate if Popup or Callback should be remembered in local storage

### Fixed
- Inextensibility error when passing arguments to signinPopup/signinRedirect

## [1.0.0] - 2018-10-12
### Added
- First release

[2.0.0]: https://github.com/fileless/ng-oidc-client/compare/v1.0.7...v2.0.0
[1.0.7]: https://github.com/fileless/ng-oidc-client/compare/v1.0.6...v1.0.7
[1.0.6]: https://github.com/fileless/ng-oidc-client/compare/v1.0.5...v1.0.6
[1.0.5]: https://github.com/fileless/ng-oidc-client/compare/v1.0.4...v1.0.5
[1.0.4]: https://github.com/fileless/ng-oidc-client/compare/v1.0.3...v1.0.4
[1.0.3]: https://github.com/fileless/ng-oidc-client/compare/v1.0.2...v1.0.3
[1.0.2]: https://github.com/fileless/ng-oidc-client/compare/v1.0.1...v1.0.2
[1.0.1]: https://github.com/fileless/ng-oidc-client/compare/v1.0.0...v1.0.1


src/app/app.module.ts
import { NgModule, APP_INITIALIZER } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { RouterModule, Routes } from '@angular/router';
import { EffectsModule } from '@ngrx/effects';
import { routerReducer, RouterReducerState } from '@ngrx/router-store';
import { ActionReducerMap, StoreModule } from '@ngrx/store';
import { StoreDevtoolsModule } from '@ngrx/store-devtools';
import { NgOidcClientModule, Config } from 'ng-oidc-client';
import { AppComponent } from './core/components/app/app.component';
import { HomeComponent } from './core/components/home/home.component';
import { OidcGuardService } from './core/providers/oidc-guard.service';
import { ProtectedComponent } from './core/components/protected/protected.component';
import { LoginComponent } from './core/components/login/login.component';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { OidcInterceptorService } from './core/providers/oidc-interceptor.service';
import { UserModule } from './modules/user/user.module';
import { WebStorageStateStore, Log } from 'oidc-client';
import { MaterialModule } from './material/material.module';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { UnauthorizedComponent } from './core/components/unauthorized/unauthorized.component';

export interface State {
  router: RouterReducerState;
}

export const rootStore: ActionReducerMap<State> = {
  router: routerReducer
};

const routes: Routes = [
  {
    path: '',
    pathMatch: 'full',
    component: HomeComponent
  },
  {
    path: 'protected',
    canActivate: [OidcGuardService],
    component: ProtectedComponent
  },
  {
    path: 'unauthorized',
    component: UnauthorizedComponent
  },
  { path: '**', redirectTo: '' }
];

export function getWebStorageStateStore() {
  return new WebStorageStateStore({ store: window.localStorage });
}

export const oidcConfigSettings: Config = {
  oidc_config: {
    authority: 'https:/ng-oidc-client-server.azurewebsites.net',
    client_id: 'ng-oidc-client-identity',
    redirect_uri: 'http://localhost:4200/callback.html',
    response_type: 'code',
    scope: 'openid profile offline_access api1',
    post_logout_redirect_uri: 'http://localhost:4200/signout-callback.html',
    silent_redirect_uri: 'http://localhost:4200/renew-callback.html',
    accessTokenExpiringNotificationTime: 10,
    automaticSilentRenew: true,
    userStore: getWebStorageStateStore
  }
  // log: {
  //   logger: console,
  //   level: 0
  // },
  // useCallbackFlag: true
};

// Oidc config on APP_INITIALIZER
// export function loadConfig(facade: OidcFacade) {
//   return () =>
//     new Promise(resolve => {
//       setTimeout(() => {
//         facade.configureOidcClient(configSettings);
//         resolve();
//       }, 2000);
//     });
// }

@NgModule({
  declarations: [AppComponent, ProtectedComponent, HomeComponent, LoginComponent, UnauthorizedComponent],
  imports: [
    BrowserModule,
    BrowserAnimationsModule,
    RouterModule.forRoot(routes),
    StoreModule.forRoot(rootStore, {
      runtimeChecks: {
        strictStateSerializability: true,
        strictActionSerializability: true,
        strictStateImmutability: true,
        strictActionImmutability: true
      }
    }),
    EffectsModule.forRoot([]),
    StoreDevtoolsModule.instrument({
      name: 'ng-oidc-client',
      logOnly: true
    }),
    NgOidcClientModule.forRoot(oidcConfigSettings),
    UserModule.forRoot({
      urls: {
        api: 'https://localhost:5001'
      }
    }),
    MaterialModule
  ],
  providers: [
    OidcGuardService,
    {
      provide: HTTP_INTERCEPTORS,
      useClass: OidcInterceptorService,
      multi: true
    }
    // {
    //   provide: APP_INITIALIZER,
    //   useFactory: loadConfig,
    //   deps: [OidcFacade],
    //   multi: true
    // }
  ],
  bootstrap: [AppComponent]
})
export class AppModule {}
